### PR TITLE
Kim mb 3716 update update shipment service

### PR DESCRIPTION
**Pull request by kimallen**
## Description

The service for updating the `mto_shipment` was prime-specific and included validations that only applied to the prime.  This PR moves all of the prime-specific validations to the primeapi `UpdateMTOShipmentHandler` and makes the Update service generic.  It also adds in the fields that are missing in the updater service.  
The service can now additionally update these fields:
- requestedPickupDate
- requestedDeliveryDate
- PrimeActualWeight
- FIrstAvailableDeliveryDate
- mtoAgents (adding new ones and updating existing ones)
-customerRemarks

This should allow the customer update mto_shipment api to use the service to update the shipment
## Reviewer Notes

[Discussion on whether services should be generic](https://ustcdp3.slack.com/archives/CP6PTUPQF/p1597880829068300)

The MTOAgents are now set up so that they can be updated if they exist, and created if they are not.  In the future they should be separated out into their own service and have an eTag applied.

The Address field updates are not currently working, and Team Roci currently has a [story](https://dp3.atlassian.net/browse/MB-3691) in their sprint to fix it.
There will also need to be added logic to create a new address field if it does not exist.

Ideally the function in the service, `CalculateRequiredDeliveryDate`, which is used by both the service and the prime handler, should have its own tests to test its functionality.  Right now the functionality is tested by the tests moved into the handler.  Another story could be written for moving the tests of that function.

## Setup

run tests for `services/mto_shipment`
run tests for `handlers/primeapi/mto_shipment`

## Code Review Verification Steps

* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-3716) for this change
[address bug story](https://dp3.atlassian.net/browse/MB-3691)
[Discussion on whether services should be generic](https://ustcdp3.slack.com/archives/CP6PTUPQF/p1597880829068300)

### CODE REVIEW

File: pkg/handlers/primeapi/mto_shipment.go [hunk 1]
```diff
@@ -215,9 +217,25 @@ func (h UpdateMTOShipmentHandler) Handle(params mtoshipmentops.UpdateMTOShipment
...
+		err := h.DB().Eager("PickupAddress", "DestinationAddress").Find(&dbShipment, params.MtoShipmentID)
+		if err != nil {
+			return mtoshipmentops.NewUpdateMTOShipmentNotFound().WithPayload(payloads.ClientError(handlers.NotFoundMessage, err.Error(), h.GetTraceID()))
+		}
+
+		mtoShipment, validationErrs := h.checkPrimeValidationsOnModel(mtoShipment, &dbShipment)
+		if validationErrs != nil && validationErrs.HasAny() {
+			logger.Error("primeapi.UpdateMTOShipmentHandler error - extra fields in request", zap.Error(validationErrs))
+
+			errPayload := payloads.ValidationError("Fields that cannot be updated found in input",
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/mto_shipment.go:230]: Oh, I think this is leftover copy pasta from the error check above! I would change this message to something like "Invalid data found in input" or something to that effect

File: pkg/handlers/primeapi/mto_shipment.go [hunk 2]
```diff
@@ -237,3 +255,78 @@ func (h UpdateMTOShipmentHandler) Handle(params mtoshipmentops.UpdateMTOShipment
...
+	if (approvedDaysFromScheduled >= 3 && approvedDaysFromScheduled <= 9) && daysFromScheduled >= 3 {
+		return nil
+	}
+
+	if approvedDaysFromScheduled < 3 && daysFromScheduled >= 1 {
+		return nil
+	}
+
+	return services.InvalidInputError{}
+}
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/mto_shipment.go:332]: Wow thank you for adding this! I feel like I never really understood what was happening with this, and this function makes it so much clearer

File: pkg/handlers/primeapi/mto_shipment_test.go [hunk 3]
```diff
@@ -532,4 +542,412 @@ func (suite *HandlerSuite) TestUpdateMTOShipmentHandler() {
...
+
+		response := handler.Handle(params)
+		suite.IsType(&mtoshipmentops.UpdateMTOShipmentOK{}, response)
+
+		okResponse := response.(*mtoshipmentops.UpdateMTOShipmentOK)
+		responsePayload := okResponse.Payload
+		suite.Equal(oldShipment.ID.String(), responsePayload.ID.String())
+		suite.NotNil(responsePayload.RequiredDeliveryDate)
+	})
+}
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/mto_shipment_test.go:953]: WOW, thank you for adding all these great tests!! 🙌

File: pkg/handlers/primeapi/mto_shipment_test.go [hunk 4]
```diff
@@ -532,4 +542,412 @@ func (suite *HandlerSuite) TestUpdateMTOShipmentHandler() {
...
+		responsePayload := okResponse.Payload
+		suite.Equal(oldShipment.ID.String(), responsePayload.ID.String())
+		suite.NotNil(responsePayload.RequiredDeliveryDate)
+
+		// Let's double check our maths.
+		expectedRDD := time.Time(responsePayload.ScheduledPickupDate).AddDate(0, 0, 12)
+		actualRDD := time.Time(responsePayload.RequiredDeliveryDate)
+		suite.Equal(expectedRDD.Year(), actualRDD.Year())
+		suite.Equal(expectedRDD.Month(), actualRDD.Month())
+		suite.Equal(expectedRDD.Day(), actualRDD.Day())
```
Comment (sandy-wright) [inline pkg/handlers/primeapi/mto_shipment_test.go:660]: Nice ++

File: pkg/services/mto_shipment/mto_shipment_updater_test.go [hunk 5]
```diff
@@ -116,238 +117,314 @@ func (suite *MTOShipmentServiceSuite) TestMTOShipmentUpdater() {
...
+	//		ID:                   oldShipment.ID,
+	//		PrimeEstimatedWeight: &primeEstimatedWeight,
+	//	}
+	//	updatedMTOShipment, err := mtoShipmentUpdater.UpdateMTOShipment(&updatedShipment, eTag)
+	//	suite.NoError(err)
+	//
+	//	suite.NotZero(updatedMTOShipment.ID, oldMTOShipment.ID)
+	//	suite.NotNil(updatedMTOShipment.PrimeEstimatedWeightRecordedDate)
+	//})
+	//
```
Comment (sandy-wright) [inline pkg/services/mto_shipment/mto_shipment_updater_test.go:427]: It looks like you moved these tests to the Prime API - can they be removed from this file?

Comment (kimallen) [reply pkg/services/mto_shipment/mto_shipment_updater_test.go:427]: Oh yes.  Thank you!

File: pkg/handlers/primeapi/mto_shipment.go (hunk 2 above)
Comment (kimallen) [reply pkg/handlers/primeapi/mto_shipment.go:332]: I can't take credit- this was an existing function- I just moved it over 🙂

**Pull request merged**